Add `ItemHandlerHelper.moveItems` helper method

Technici4n opened the pull request:
This PR adds a new helper method to move items from one inventory to another. This is surprisingly tricky to get right, and should be provided to all users of NeoForge for convenience and correctness.

Typical use cases include machine auto-import/export, pipes moving items, etc...

Technici4n committed:
Add `ItemHandlerHelper.moveItems` helper method

shartte commented on src/main/java/net/neoforged/neoforge/items/ItemHandlerHelper.java:
```diff
@@ -185,4 +191,84 @@ public static int calcRedstoneFromInventory(@Nullable IItemHandler inv) {
+            Predicate<ItemStack> filter,
+            int maxAmount,
+            boolean stackInTarget) {
+        Objects.requireNonNull(filter);
```
Dunno how consistently we do this, but IMHO in APIs we should provide the name of the argument that was null...

```suggestion
        Objects.requireNonNull(filter, "filter");
```

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/items/ItemHandlerHelper.java:
```diff
@@ -185,4 +191,84 @@ public static int calcRedstoneFromInventory(@Nullable IItemHandler inv) {
+                    }
+                }
+
+                if (movedThisTime <= 0) {
+                    break;
+                }
```
Since this is the only `break` in the loop, consider moving `movedThisTime` to outside the loop and turning this loop into a `do {} while ();` instead.

Technici4n replied:
I prefer not to, as it increases the scope of the variable unnecessarily, and doesn't particularly improve readability IMO. :D

sciwhiz12 commented on src/main/java/net/neoforged/neoforge/items/ItemHandlerHelper.java:
```diff
@@ -185,4 +191,84 @@ public static int calcRedstoneFromInventory(@Nullable IItemHandler inv) {
+                    leftover = from.insertItem(i, leftover, false);
+
+                    if (!leftover.isEmpty()) {
+                        LOGGER.warn(
+                                "Trying to move up to {} items from {} to {}, but destination rejected {} that could not be inserted back in the source.",
+                                maxAmount, from, to, leftover);
```
Perhaps we should add additional logging in development (via `FMLEnvironment.production`) for the stacktrace in question, to aid developers in hunting down where this warning is coming from.

Technici4n replied:
done

sciwhiz12 reviewed (changes requested):
Hello! Do you plan on continuing this PR?

Technici4n committed:
review

Technici4n commented:
Since there has been some discussion on making the `IItemHandler` experience nicer, I addressed the reviews and updated this PR to 1.21. 😉

To clarify why this is helpful, here are some variations of this code in multiple places:
- https://github.com/AztechMC/Modern-Industrialization/blob/ff98121446620064c98659c5d9395ea241075ba5/src/main/java/aztech/modern_industrialization/util/TransferHelper.java#L33
- https://github.com/Technici4n/Modern-Dynamics/blob/c3d58bdab24df0507d5194433c77af90052b79f4/src/main/java/dev/technici4n/moderndynamics/network/item/ItemHost.java#L278
- A lot of common item moving logic (for example auto-import, auto-export) can be changed to use it as well.

Technici4n committed:
break

HenryLoenwind commented on src/main/java/net/neoforged/neoforge/items/ItemHandlerHelper.java:
```diff
@@ -146,4 +154,89 @@ public static int calcRedstoneFromInventory(@Nullable IItemHandler inv) {
+            @Nullable IItemHandler from,
+            @Nullable IItemHandler to,
+            Predicate<ItemStack> filter,
+            int maxAmount,
```
maxAmount is never used...

HenryLoenwind commented on src/main/java/net/neoforged/neoforge/items/ItemHandlerHelper.java:
```diff
@@ -146,4 +154,89 @@ public static int calcRedstoneFromInventory(@Nullable IItemHandler inv) {
+
+                // Simulate insertion
+                var simulationLeftover = stackInTarget ? insertItemStacked(to, available, true) : insertItem(to, available, true);
+                int canFit = availableCount - simulationLeftover.getCount();
```
From experience, this number will be extremely unreliable. Simulating inserting items into multiple slots is a bad idea, there are many targets where the slots are not independent. You need a copy of insertItemStacked that returns the number it can insert into the first target it can insert into. Then execute that insert and repeat until maxItems or source empty.

One simple example would be the Ender IO Alloy Smelter. It has 3 input slots, and if you use insertItemStacked with Iron Ingots, it will report that it can insert 192 items, i.e. a full stack in each slot. But, as soon as there's iron in one slot, it will lock in a recipe and the other two slots won't accept any iron.

Here's some proven code: https://github.com/SleepyTrousers/EnderIO-1.5-1.12/blob/master/enderio-base/src/main/java/crazypants/enderio/base/capability/ItemTools.java#L370 (Line 170 has the moveItems() equivalent, although it's a bit less readable as it has a couple extra features). And even with that we had spilled items on occasion.

PS: Feel free to copy as much from there as you want, there is exactly one non-empty line that is not fully by me.

Technici4n replied:
Yes you are correct. Choosing how to handle these cases is the difficult aspect of this PR, and it highlights how difficult `IItemHandler` is to compose with higher-level abstractions. (`insertItemStacked` is kinda broken)

HenryLoenwind replied:
It works sufficiently well if you stop after the first successful simulated operation. Because that should be repeatable and work for a non-simulated identical operation that is executed in the same context.

There still are one logic issue and one performance issue:

- The actual operation must be identical, i.e. the itemstack size needs to be the same as in the simulation. A smaller size may lead to different results (it's rare, but only being able to hold, e.g., integer multiples of recipe amounts is legal). So, using the simulation to determine the maximum insertable size and then min() it with other constraints doesn't work.
- Running the simulation over the target slot by slot to find one where it can insert, and then doing the same thing for the actual insert wastes CPU (remember that Giant Chests debacle?). The slot number of the slot where the simulation succeeded should be reused.

Note that the code I linked is guilty of both. Those optimisations are in the conduit code, but that code isn't readable outside that context. I never had a need to put them into the machine pull/push code, as they're not that critical there.

---

Warstory Mode: Giant Chests (I think that was what the mod was called) got included in a popular modpack (I think it was one of the skyblock ones) and got featured by basically every YouTuber playing that pack. However, it quickly turned out that there was not a single item transport solution that wouldn't lag the server badly trying to insert into its flat inventory with thousands of slots. Server operators quickly banned that mod and (some) item transport mods slowly implemented slot scanning limits. But those require the pipe to remember where it stopped scanning for insertable slots in the previous tick, making the whole logic even more complicated. If you ever wondered why an Item Conduit keeps adding to the same stack, even if you manually empty a slot before that one, something a Hopper does not, this is why.

HenryLoenwind commented:
> This is surprisingly tricky to get right, 

quod erat demonstrandum...

shartte commented:
Tests? :-)

embeddedt commented:
@Technici4n Do you still plan to update this?

Technici4n commented:
Not sure, there are no good solutions here and this PR just highlights how flawed IItemHandler is.

embeddedt commented:
Can we close it for now then and revive it later?
